### ISSUE TITLE
I' suggest using determinism by default making predictions

### ISSUE THREAD

**Issue by 3niQma** [hill-a/stable-baselines#819]
Even though I spend a lot of time with sb, I didn't realize that my predictions of an trained models is not by default deterministic but samples from the policy's probability distribution. Isn't it rare that you want to apply a model without always choosing what maximizes your policy distribution? Why then having determinism set off by default in the prediction function? 

In one example I am working on, variance of the gaussian distribution seems to be still large after training reducing my model's performance by ~30% if not using deterministic predictions.

So my suggestion is to either use by default **deterministic=True** in the predict method of **class BaseRLModel** (https://github.com/hill-a/stable-baselines/blob/master/stable_baselines/common/base_class.py). 

If that is unwanted, you might consider changing examples (e.g. PPO cartpole) as follows:

obs = env.reset()
while True:
    action, _states = model.predict(obs, **deterministic=True**)
    obs, rewards, dones, info = env.step(action)
    env.render()

**Comment by Miffyli** [comment]
I see the motivation behind this, but I think we should stick with default stochastic actions as this is what all RL work uses by default, unless otherwise noted. Using deterministic actions can sometimes lead agent to states where it gets stuck, and from theory-point-of-view there are scenarios where stochastic actions are necessary for solving the environment (I do not have sources to cite at hand right now, sadly).

**Comment by 3niQma** [comment]
I see what you mean. It's mentioned in the _Sutton_ book _Policy Gradient_ chapter when the authors note that some optimal policies are stochastic (like randomly bluffing at poker from time to time.

**Comment by araffin** [comment]
>I see the motivation behind this, but I think we should stick with default stochastic actions as this is what all RL work uses by default, unless otherwise noted. Using deterministic actions can sometimes lead agent to states where it gets stuck, and from theory-point-of-view there are scenarios where stochastic actions are necessary for solving the environment (I do not have sources to cite at hand right now, sadly).

totally agree, that's why it is also the default behavior for most algorithms.
This is explicitely [mentioned](https://stable-baselines.readthedocs.io/en/master/guide/rl_tips.html) in the doc:

> As some policy are stochastic by default (e.g. A2C or PPO), you should also try to set deterministic=True when calling the .predict() method, this frequently leads to better performance. Looking at the training curve (episode reward function of the timesteps) is a good proxy but underestimates the agent true performance.


>Above was mainly for discrete actions. I will let @araffin comment on continuous actions as he is more familiar with those (I would be curious to know if deterministic actions are common there or not).

For continuous actions, you don't want the final policy to be stochastic (as you are usually using robots, a small mistake can result in a total failure). That's why SAC/DDPG/TD3 have `deterministic=True` by default when you call the predict method.

**Comment by 3niQma** [comment]
I see. Could you elaborate why it is `deterministic=True` for SAC/DDPG/TD3 but not for PPO?

**Comment by araffin** [comment]
For DDPG/TD3, it is because the underlying policy is deterministic (please read the papers).
For SAC, it is something that is recommended in the paper (again) to have better performance at test time and which makes sense for continuous actions. (I'm also writing a paper that should be publish soon which give you another reason why `deterministic=False` is a bad idea for continuous actions)

PPO can be used both for discrete and continuous actions (which is not the case for the other algos), and as the default hyperparameters are for discrete actions, we chose to keep it stochastic. However, it is recommended (as mentioned in the doc) to use deterministic actions at test time when using continuous actions.

**Comment by 3niQma** [comment]
Okay, great. Everything is clarified then. Thank you so much for your effort also for the library in general. BR
